### PR TITLE
[SR-1052] Use @discardableResult and warn by default

### PR DESCRIPTION
**Pull request by tanadeau**
#### What's in this pull request?

Part 5 of SR-1052: Warnings for unused results are now turned on by default. These can be silenced with the `@discardableResult` attribute. The old `@warn_unused_result` is now ignored and will be removed in a future PR.
#### Resolved bug number: ([Part 5 of SR-1052](https://bugs.swift.org/browse/SR-1052))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by tanadeau** [comment]
CC: @jckarter since he helped me to get the warning working with functions with a generic return type.

File: stdlib/private/SwiftPrivatePthreadExtras/SwiftPrivatePthreadExtras.swift [hunk 1]
```diff
@@ -125,7 +125,7 @@ public class _stdlib_Barrier {
   }
 
   deinit {
-    _stdlib_pthread_barrier_destroy(_pthreadBarrierPtr)
+    _ = _stdlib_pthread_barrier_destroy(_pthreadBarrierPtr)
```
Comment (gribozavr) [inline stdlib/private/SwiftPrivatePthreadExtras/SwiftPrivatePthreadExtras.swift]: I think this should be:

```swift
  let ret = _stdlib_pthread_barrier_destroy(_pthreadBarrierPtr)
  if ret != 0 {
    fatalError("_stdlib_pthread_barrier_init() failed")
  }
```

File: validation-test/stdlib/CollectionDiagnostics.swift [hunk 2]
```diff
@@ -58,11 +58,11 @@ func sortResultIgnored<
...
 
-  mutableCollection.sorted() // expected-warning {{result of call to non-mutating function 'sorted()' is unused; use 'sort()' to mutate in-place}} {{21-27=sort}}
-  mutableCollection.sorted { $0 < $1 } // expected-warning {{result of call to non-mutating function 'sorted(isOrderedBefore:)' is unused; use 'sort(isOrderedBefore:)' to mutate in-place}} {{21-27=sort}}
+  mutableCollection.sorted() // expected-warning {{result of call to 'sorted()' is unused}}
+  mutableCollection.sorted { $0 < $1 } // expected-warning {{result of call to 'sorted(isOrderedBefore:)' is unused}}
 
-  array.sorted() // expected-warning {{result of call to non-mutating function 'sorted()' is unused; use 'sort()' to mutate in-place}} {{9-15=sort}}
-  array.sorted { $0 < $1 } // expected-warning {{result of call to non-mutating function 'sorted(isOrderedBefore:)' is unused; use 'sort(isOrderedBefore:)' to mutate in-place}} {{9-15=sort}}
+  array.sorted() // expected-warning {{result of call to 'sorted()' is unused}}
+  array.sorted { $0 < $1 } // expected-warning {{result of call to 'sorted(isOrderedBefore:)' is unused}}
```
Comment (gribozavr) [inline validation-test/stdlib/CollectionDiagnostics.swift]: Did we lose the fixit?

Comment (tanadeau) [inline validation-test/stdlib/CollectionDiagnostics.swift]: Yes. That was noted in the proposal (SE-0047) as a downside. To partially make up for this, there are now `mutatingVariant` and `nonmutatingVariant` doc tags that interact with code completion. I added those in #2204.

File: stdlib/private/SwiftPrivatePthreadExtras/SwiftPrivatePthreadExtras.swift (hunk 1 above)
Comment (tanadeau) [inline stdlib/private/SwiftPrivatePthreadExtras/SwiftPrivatePthreadExtras.swift]: Thanks! This is exactly the kind of feedback I was hoping for. I wasn't sure what returns were intentionally ignored and which were accidental.

Comment (tanadeau) [inline stdlib/private/SwiftPrivatePthreadExtras/SwiftPrivatePthreadExtras.swift]: Fixed.

**Comment by natecook1000** [comment]
`SetAlgebra`, `OptionSet`, and `Set` all have `update` and `remove` methods that could use the `@discardableResult` attribute, as well. (It looks like you already got `Set.remove`.)

**Comment by tanadeau** [comment]
@natecook1000 I rebased against master and added the missing `@discardableResult` attributes to the `update` and `remove` methods.

**Comment by tanadeau** [comment]
I rebased again and resolved the conflicts caused by the new floating point protocols.

File: include/swift/AST/DiagnosticsSema.def [hunk 3]
```diff
@@ -2196,11 +2196,6 @@ WARNING(expression_unused_result,none,
         "result of call to %0 is unused", (DeclName))
 WARNING(expression_unused_init_result,none,
         "result of initializer is unused", ())
-WARNING(expression_unused_result_message,none,
-        "result of call to %0 is unused: %1", (DeclName, StringRef))
-WARNING(expression_unused_result_nonmutating,none,
-        "result of call to non-mutating function %0 is unused; "
-        "use %1 to mutate in-place", (DeclName, DeclName))
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: I think losing this diagnostic is pretty unfortunate.

File: test/Parse/consecutive_statements.swift [hunk 4]
```diff
@@ -7,7 +7,7 @@ func statement_starts() {
   f(0)
   f (0)
   f // expected-error{{expression resolves to an unused l-value}}
-  (0)
+  (0) // expected-warning {{result of call to 'init(_builtinIntegerLiteral:)' is unused}}
```
Comment (jrose-apple) [inline test/Parse/consecutive_statements.swift]: Ick. We should offer a better diagnostic for implicit calls. That can come later, though.

File: include/swift/AST/DiagnosticsSema.def (hunk 3 above)
Comment (tanadeau) [inline include/swift/AST/DiagnosticsSema.def]: That was discussed as part of the proposal: https://github.com/apple/swift-evolution/blob/master/proposals/0047-nonvoid-warn.md#mutating-variants

**Comment by jrose-apple** [comment]
Sema change looks good to me.

File: include/swift/AST/DiagnosticsSema.def (hunk 3 above)
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: Yeah, I saw after I added the comment. Thanks.

File: test/Parse/consecutive_statements.swift (hunk 4 above)
Comment (tanadeau) [inline test/Parse/consecutive_statements.swift]: Agreed on both counts. These changes did fix an issue where unused literals didn't create any warning.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by tanadeau** [comment]
There was a failure in the validation tests under tvOS that appears to be completely unrelated:

```
Command Output (stderr):
--
/Users/buildnode/jenkins/workspace/swift-PR-osx/swift/validation-test/Reflection/functions.swift:117:14: error: expected string not found in input
// CHECK-64: Type info:
             ^
<stdin>:51:1: note: scanning from here
Child process terminated with signal 13: Broken pipe
^
<stdin>:51:13: note: possible intended match here
Child process terminated with signal 13: Broken pipe
```

**Comment by gribozavr** [comment]
The failure looks unrelated.

**Comment by lattner** [comment]
This is so great, congrats @tanadeau!!

**Pull request merged**